### ISSUE TITLE
[k8s] Fractional GPU clusters consume full GPU

### ISSUE THREAD

**Issue by romilbhardwaj** [skypilot-org/skypilot#2655]
From @Michaelvll:

> 1. `sky launch --gpus V100:0.5 --num-nodes 3 "hostname -i"`
> 2. `sky status`: the resources column shows `3x Kubernetes(2CPU--8GB--1V100, {'V100': 1})`, instead of `V100:0.5`, which is a bit surprising.

Unlike CPU, GPU requests cannot be fractional in Kubernetes:
```
# Trying to create a deployment requesting nvidia.com/gpu: 0.5
* spec.containers[0].resources.limits[nvidia.com/gpu]: Invalid value: resource.Quantity{i:resource.int64Amount{value:5, scale:-1}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"", Format:"DecimalSI"}: must be an integer
```

Need to look into whether we should disallow fractional GPU requests when launching (note that ray can still handle fractional resources, so fractional requests at `sky exec` _may_ still work (TBD)).

**Comment by romilbhardwaj** [comment]
Reopening - this can potentially be confusing behavior for users.

**Comment by bryantbiggs** [comment]
devices in Kubernetes are consumed through the [extended resources API ](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/) which only allows integer values

> Extended resources must be advertised in integer amounts. For example, a Node can advertise four dongles, but not 4.5 dongles

NVIDIA does support various means for [concurrency models](https://developer.nvidia.com/blog/improving-gpu-utilization-in-kubernetes/), but this means that you need to also control these concurrency/sharing controls. Unfortunately, using the extended resources API to define resource requirements is quite crude currently - instead, a new KEP has been put forward to improve support for devices and their resource request/allocation:
- [Dynamic Resource Allocation](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3063-dynamic-resource-allocation/README.md)
- [Dynamic Resource Allocation KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3063-dynamic-resource-allocation/README.md)

In short - it will be quite challenging for skypilot to add support for fractional GPUs since its currently highly dependent on cluster administrators providing the necessary components (i.e. NVIDIA K8s device plugin) and their associated configurations (configmap for time-slicing or MiG, etc), and then skypilot would need to check for these resources, get their configuration values, and be able to translate the requested value (i.e. - `--gpus V100:0.5`) into a suitable integer value based on the prior information collected from the cluster.

**Comment by romilbhardwaj** [comment]
Thanks for insights @bryantbiggs! Agreed, until k8s allows fractional extended resources, our best bet is to translate fractional GPU requests from SkyPilot to integer requests to k8s (which are again, dependent on how MIG is setup on the cluster).

Is this feature of interest to you? If you have a specific use case in mind, happy to brainstorm more ideas here!

**Comment by bryantbiggs** [comment]
> Is this feature of interest to you? If you have a specific use case in mind, happy to brainstorm more ideas here!

Its an area of interest as I help people run/deploy LLM workloads on Kubernetes, but nothing specific here other than the issue was pointed out by a user waiting for fractional GPU support. Just thought I'd clarify a bit to save any future readers

**Comment by dominiquegarmier** [comment]
Are there any developments on this? I.e. does kubermetes now support this?
